datastore/longhorn.go

```diff
@@ -233,6 +232,71 @@ func (s *DataStore) ListSettings() (map[types.SettingName]*longhorn.Setting, err
 	return itemMap, nil
 }
 
+func (s *DataStore) annotateAWSIAMRole(awsIAMRole string) error {
+	addAWSIAMRole := awsIAMRole != ""
+
+	// create/update/delete AWS IAM Role annotation to all longhorn-manager Pods
+	selector, err := s.getManagerSelector()
+	if err != nil {
+		return err
+	}
+	podList, err := s.pLister.Pods(s.namespace).List(selector)
+	if err != nil {
+		return err
+	}
+	for _, pod := range podList {
+		val, exist := pod.Annotations[types.AWSIAMRroleAnnotation]
+
+		if addAWSIAMRole && (!exist || val != awsIAMRole) {
+			// add annotation to the longhorn-manager Pods if the annotation is not present or the annotation needs to be updated
+			pod.Annotations[types.AWSIAMRroleAnnotation] = awsIAMRole
+			_, err = s.kubeClient.CoreV1().Pods(s.namespace).Update(pod)
+			if err != nil {
+				return err
+			}
+		} else if !addAWSIAMRole && exist {
+			// delete annotation to the longhorn-manager Pods if the annotation is present
+			delete(pod.Annotations, types.AWSIAMRroleAnnotation)
+			_, err = s.kubeClient.CoreV1().Pods(s.namespace).Update(pod)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// create/update/delete AWS IAM Role annotation to all instance-manager-r Pods
+	selector, err = metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: types.GetInstanceManagerLabels("", "", types.InstanceManagerTypeReplica),
+	})
+	if err != nil {
+		return err
+	}
+	podList, err = s.pLister.Pods(s.namespace).List(selector)
+	if err != nil {
+		return err
+	}
+	for _, pod := range podList {
+		val, exist := pod.Annotations[types.AWSIAMRroleAnnotation]
+
+		if addAWSIAMRole && (!exist || val != awsIAMRole) {
+			// add annotation to the instance-manager-r Pods if the annotation is not present or the annotation needs to be updated
+			pod.Annotations[types.AWSIAMRroleAnnotation] = awsIAMRole
+			_, err = s.kubeClient.CoreV1().Pods(s.namespace).Update(pod)
+			if err != nil {
+				return err
+			}
+		} else if !addAWSIAMRole && exist {
+			// delete annotation to the instance-manager-r Pods if the annotation is present
+			delete(pod.Annotations, types.AWSIAMRroleAnnotation)
+			_, err = s.kubeClient.CoreV1().Pods(s.namespace).Update(pod)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
```

Comment:
```suggestion
// annotateAWSIAMRole ensures that the running pods of the manager as well as the replica instance managers.
// have the correct IAM role assigned to them based on the passed `awsIAMRole`
func (s *DataStore) annotateAWSIAMRole(awsIAMRole string) error {
	selector, err := s.getManagerSelector()
	if err != nil {
		return err
	}
	
	managerPods, err := s.pLister.Pods(s.namespace).List(selector)
	if err != nil {
		return err
	}
	
	selector, err = metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
		MatchLabels: types.GetInstanceManagerLabels("", "", types.InstanceManagerTypeReplica),
	})
	if err != nil {
		return err
	}

	imPods, err := s.pLister.Pods(s.namespace).List(selector)
	if err != nil {
		return err
	}
	
	pods = append(managerPods, imPods)
	for _, pod := range pods {
		val, exist:= pod.Annotations[types.AWSIAMRroleAnnotation]
		updateAnnotation := awsIAMRole != "" && awsIAMRole != val
		deleteAnnotation := awsIAMRole == "" && exist

		if updateAnnotation  {
			pod.Annotations[types.AWSIAMRroleAnnotation] = awsIAMRole
		} else if deleteAnnotation {
			delete(pod.Annotations, types.AWSIAMRroleAnnotation)
		} else {
			continue
		}
		
		if _, err = s.kubeClient.CoreV1().Pods(s.namespace).Update(pod); err != nil {
			return err
		}
	}

	return nil
}
```

Refactor to much code duplication, by restructuring the code to the above you can easily follow the logic and add trace logs if desired :)

Reply:
Thank u, really cool!